### PR TITLE
Stats Widget: Load stats in the widget after user logs in

### LINKED ISSUE #17567: Stats widget not working until user visits Stats screen in app
### Expected behavior

If a user adds a WordPress widget to their device's home screen, the widget should work out of the box or display an actionable message to help the user get it functioning.

### Actual behavior

After adding the widget, it shows "Unable to load site stats". This happens only if the user hasn't visited the app's Stats screen after logging in.

### Steps to reproduce the behavior

1. Perform a fresh install of WPiOS
2. Log in to the app
3. Background the app and add the Today widget
4. Notice the "Unable to load site stats" message appears, with no instruction on how to remedy this
5. Notice that tapping on the widget opens the app, but does not deep link to the Stats screen (see https://github.com/wordpress-mobile/WordPress-iOS/issues/18452)

### Possible solution

My understanding is that there is an implicit requirement that the user visits the Stats page in the app for widgets to work. If widgets must depend on that user action taking place first, the widget could show the actionable message instead of "Unable to load site stats". For example, we could use "Visit the app's Stats page to enable the widget".

##### Tested on iPhone 11, iOS 15.0.2, WPiOS 18.6.1

First reported by @AliSoftware

### PR DESCRIPTION
**Pull request by staskus**
## Issue

Closes #17567 

### Issue

Data on the stats widget is not shown until the user manually visits the stats page.

### Rootcause

There are 2 reasons:
1. App only calls stats widget initialization in `StatsViewController.initStats()`
2. Stat widget itself in `SiteListProvider.getTimeline(...)` doesn't fetch stats while `configuration.site` is nil

### Solution

1. Listen to `WPSigninDidFinishNotification` and initialize stats when the event is received. 
Two important pieces of data are already loaded when this event triggers: 1) `WPStatsHomeWidgetsUserDefaultsSiteIdKey` is set 2) `visibleBlogsForWPComAccounts` are set

2. Update `SiteListProvider` so it would use `defaultSiteID` while `configuration.site` is nil. This allows Stats widget to load required data itself in `SiteListProvider.getTimeline(...)` and not rely on `StatsViewController` being opened for data to be loaded

### Another potential improvement

Create an empty state for the widget while the data is not loaded. For example show the site name and "-" instead of numbers.

## Testing instructions

1st scenario:

1. Open the app and make sure it's logged out
2. Add Widgets
4. Log in
5. Background the app
6. Confirm that Stats load

2nd scenario:

1. Open the app and make sure it's logged out
4. Log in
5. Background the app
7. Add Widgets
8. Confirm that Stats load

3rd scenario:
1. Open the app and make sure it's logged out
2. Log in with the account
3. Add widget
4. Long tap on the widget and click Edit Widget
5. Select a site
6. Come back to the app
7. Logout
8. Log in with the different account
9. Background the app
10. Confirm the Stats load

## Regression Notes

1. Potential unintended areas of impact

Make sure the widget is still updated when visiting Stats view

2. What I did to test those areas of impact (or what existing automated tests I relied on)

3. What automated tests I added (or what prevented me from doing so)

Added `StatsWidgetsStore` to confirm that Stats widget data is loaded or deleted depending on notifications posted. More thorough tests would require a refactoring of `StatsWidgetsStore` to inject WidgetCenter, AccountHelper, HomeWidgetCache, and other dependencies.

## Video

Add widgets before logging in:

https://user-images.githubusercontent.com/4062343/174248260-740ce27d-a9e6-4aa6-a114-60e1b56f048d.mov

Add widgets after logging in:

https://user-images.githubusercontent.com/4062343/173892420-e5ded3ce-09d3-43c3-a60c-c4fdb2b887f2.mov

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by staskus** [comment]
Hey @Gio2018, I saw you worked on this Widget code before. Maybe you could take a look at this PR? 👀

CC: @guarani

File: WordPress/Classes/Stores/StatsWidgetsStore.swift [hunk 1]
```diff
@@ -264,6 +265,10 @@ extension StatsWidgetsStore {
 // MARK: - Login/Logout notifications
 private extension StatsWidgetsStore {
 
+    /**
+     When WPAccountDefaultWordPressComAccountChanged is posted after login
+     the site data is not yet loaded and cannot be cached for widgets.
+     */
```
Comment (Gio2018) [inline WordPress/Classes/Stores/StatsWidgetsStore.swift]: any chance we could use modern-style documentation for methods, using the typical format, like so?

```
/// Store stats in the widget cache
    /// - Parameters:
    ///   - widgetType: concrete type of the widget
    ///   - stats: stats to be stored
```

File: WordPress/Classes/Stores/StatsWidgetsStore.swift [hunk 2]
```diff
@@ -286,6 +291,22 @@ private extension StatsWidgetsStore {
             WidgetCenter.shared.reloadTimelines(ofKind: WPHomeWidgetAllTimeKind)
         }
     }
+
+    /**
+     When WPSigninDidFinishNotification is posted after login
+     the site data data is loaded and widget data can be cached.
+     */
```
Comment (Gio2018) [inline WordPress/Classes/Stores/StatsWidgetsStore.swift]: same observation for documentation here

File: WordPress/Classes/Stores/StatsInsightsStore.swift [hunk 3]
```diff
@@ -1205,105 +1205,3 @@ private extension InsightStoreState {
         return true
     }
 }
-
-// MARK: - iOS 14 Widgets Data
-private extension InsightStoreState {
```
Comment (Gio2018) [inline WordPress/Classes/Stores/StatsInsightsStore.swift:1210]: good catch on spotting this leftover!

**Review by Gio2018 — CHANGES_REQUESTED** [review]
Hello @staskus ! 

I left some minor code comments, but other than that, I was testing the first scenario and noticed that, once logged in, the widget still does not display stats until I visit the Stats page. Once I do, it does, but the title is set to a different site (This might have something to do with some weird caching that Widgets do under the hood)
Am I missing something? Here's a brief animation of what I saw:

<p align=center>

</p>

**Comment by staskus** [comment]
Thanks for the comments and the review! @Gio2018 

There are 2 cases, the first one is not expected, and the second one is:

> I was testing the first scenario and noticed that, once logged in, the widget still does not display stats until I visit the Stats page.
>

1. Yes, you're not missing anything, you shouldn't need to visit "Stats" page for Widget to load. It could be related to caching as you mentioned. After login `reloadTimelines` should trigger the reload of Widgets. The fact that "Log in" error is shown suggests that not even "No data" state is loaded once Widget knows that the user is logged in.

I'll try to understand this issue better.

> Once I do, it does, but the title is set to a different site.
>

2. The site is always set to a default site unless it's changed in the "Edit Widget" menu. It allows having multiple widgets for different sites at the same time.

image

**Comment by staskus** [comment]
@Gio2018 

I identified the case when similar loading issue happens if:
1. Log in with account A
2. Configure widgets
3. Log in with account B
4. Widget doesn't load

Could it explain what happened in your testing case?

I found the issue with cached information of `SelectSiteIntent` configuration. If site is selected in the widget and the user logs out, `Configuration.site.identifier` still exists. [I haven't found a way to remove this information from code.](https://stackoverflow.com/questions/63764083/how-to-reset-intent-extension-configurations-in-widgetkit)

I improved `getTimelines` method to handle this case and added comments to explain such caching case. 

## Testing

Added Testing scenario 3 to the description

## Videos

Before the change:

https://user-images.githubusercontent.com/4062343/174075611-54c26bf2-6190-4ff4-9566-d7e92e70358d.mp4

After the change:

https://user-images.githubusercontent.com/4062343/174075675-ce647087-affa-4565-a18e-28d49021a3e2.mp4

**Review by Gio2018 — APPROVED** [review]
Awesome @staskus , it worked this time! Tested both scenarios.
There is a bit of delay to load stats in the widget after login, meaning that when I log in and background the app, I see "Unable to load stats" for a moment, and then the widget is populated.
Not sure if this is normal or there is something that could be improved in the way we load data (I was testing this on a fast connection so I would expect that data would load fast), but this is certainly something that can be tackled in a separate PR.

**Comment by staskus** [comment]
> Awesome @staskus , it worked this time! Tested both scenarios.
> There is a bit of delay to load stats in the widget after login, meaning that when I log in and background the app, I see "Unable to load stats" for a moment, and then the widget is populated.
> Not sure if this is normal or there is something that could be improved in the way we load data (I was testing this on a fast connection so I would expect that data would load fast), but this is certainly something that can be tackled in a separate PR.

Yes, thanks for the review again, @Gio2018! 

I had a wrong impression this delay happens due to `fetchStats` call. However, it actually happens due to excessive calls to `reloadTimelines` after login:
- We call it once after `WPAccountDefaultWordPressComAccountChanged`
- And almost immediately we call it after `WPSigninDidFinishNotification`

What we need to do is:
- `reloadTimelines` after `WPAccountDefaultWordPressComAccountChanged` when user logged out
- `reloadTimelines` after `WPSigninDidFinishNotification` when user logged in

It solved delay issue, the UX is much better now:

https://user-images.githubusercontent.com/4062343/174248260-740ce27d-a9e6-4aa6-a114-60e1b56f048d.mov

**Comment by guarani** [comment]
Thanks for the PR, @staskus! I'm merging this now given that the commits added after the last PR review approval have been discussed and informally approved.

cc @Gio2018

**Pull request merged**